zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/recovery/FileSystemRecoveryStorage.java

```diff
@@ -0,0 +1,131 @@
+import org.apache.hadoop.fs.PathFilter;
+import org.apache.hadoop.io.IOUtils;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.zeppelin.conf.ZeppelinConfiguration;
+import org.apache.zeppelin.interpreter.InterpreterSetting;
+import org.apache.zeppelin.interpreter.InterpreterSettingManager;
+import org.apache.zeppelin.interpreter.ManagedInterpreterGroup;
+import org.apache.zeppelin.interpreter.launcher.InterpreterClient;
+import org.apache.zeppelin.interpreter.remote.RemoteInterpreterProcess;
+import org.apache.zeppelin.interpreter.remote.RemoteInterpreterRunningProcess;
+import org.apache.zeppelin.notebook.FileSystemStorage;
+import org.apache.zeppelin.notebook.repo.FileSystemNotebookRepo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
+/**
+ * Hadoop compatible FileSystem based RecoveryStorage implementation.
+ *
+ * Save InterpreterProcess in the format of:
+ * InterpreterGroupId host:port
+ */
+public class FileSystemRecoveryStorage extends RecoveryStorage {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(FileSystemRecoveryStorage.class);
+
+  private InterpreterSettingManager interpreterSettingManager;
+  private ZeppelinConfiguration zConf;
+  private FileSystemStorage fs;
+  private Path recoveryDir;
+
+  public FileSystemRecoveryStorage(ZeppelinConfiguration zConf,
+                                   InterpreterSettingManager interpreterSettingManager)
+      throws IOException {
+
+    this.interpreterSettingManager = interpreterSettingManager;
+    this.zConf = zConf;
+    this.fs = FileSystemStorage.get(zConf);
+    this.recoveryDir = this.fs.makeQualified(new Path(zConf.getNotebookDir()));
+    LOGGER.info("Using folder {} to store recovery data", recoveryDir);
+    this.fs.tryMkDir(recoveryDir);
+  }
+
+  @Override
+  public void onInterpreterClientStart(InterpreterClient client) throws IOException {
+    save(client.getInterpreterSettingName());
+  }
+
+  @Override
+  public void onInterpreterClientStop(InterpreterClient client) throws IOException {
+    save(client.getInterpreterSettingName());
+  }
+
+  private void save(String interpreterSettingName) throws IOException {
+    InterpreterSetting interpreterSetting =
+        interpreterSettingManager.getInterpreterSettingByName(interpreterSettingName);
+    List<String> recoveryContent = new ArrayList<>();
+    for (ManagedInterpreterGroup interpreterGroup : interpreterSetting.getAllInterpreterGroups()) {
+      RemoteInterpreterProcess interpreterProcess = interpreterGroup.getInterpreterProcess();
+      if (interpreterProcess != null) {
+        recoveryContent.add(interpreterGroup.getId() + "\t" + interpreterProcess.getHost() + ":" +
```

Comment:
what's the `\t` for?

Reply:
how much of this kind of custom format is needed? should we look into a well known format instead?
(I guess this is a CSV/TSV file then)

Reply:
yes, it is tsv file, `\t` is used as delimited. The format is `interpreterGroupId \t host:port`